source/llassetgen/source/Packing.cpp

```diff
@@ -0,0 +1,88 @@
+#include <llassetgen/Packing.h>
+
+#include <algorithm>
+#include <tuple>
+
+namespace llassetgen {
+    Packing::Packing(Vec2<PackingSizeType> _atlasSize) : atlasSize{_atlasSize} {}
+
+    namespace impl_packing {
+        bool ShelfNextFitPacker::packNext(Vec2<PackingSizeType> rectSize) {
+            if (currentShelfSize.x + rectSize.x > packing.atlasSize.x) {
+                openNewShelf();
+                if (rectSize.x > packing.atlasSize.x) {
+                    return false;
+                }
+            }
+
+            if (usedHeight + rectSize.y > packing.atlasSize.y) {
+                return false;
+            }
+
+            store(rectSize);
+            return true;
+        }
+
+        bool ShelfNextFitPacker::packNextRotatable(Vec2<PackingSizeType> rectSize) {
+            PackingSizeType minSide, maxSide;
+            Vec2<PackingSizeType> rotated;
+            std::tie(minSide, maxSide) = std::minmax(rectSize.x, rectSize.y);
+            PackingSizeType remainingWidth = packing.atlasSize.x - currentShelfSize.x;
+            PackingSizeType remainingHeight = packing.atlasSize.y - usedHeight;
+
+            if (currentShelfSize.y >= maxSide && remainingWidth >= minSide) {
+                rotated = {minSide, maxSide};
+            } else if (remainingWidth >= maxSide && remainingHeight >= minSide) {
+                rotated = {maxSide, minSide};
+            } else {
+                openNewShelf();
+                if (usedHeight + minSide > packing.atlasSize.y) {
+                    return false;
+                }
+
+                if (maxSide > packing.atlasSize.x) {
+                    if (usedHeight + maxSide > packing.atlasSize.x) {
+                        return false;
+                    }
+
+                    rotated = {minSide, maxSide};
+                } else {
+                    rotated = {maxSide, minSide};
+                }
+            }
+
+            store(rotated);
+            return true;
+        }
+
+        void ShelfNextFitPacker::openNewShelf() {
+            usedHeight += currentShelfSize.y;
+            currentShelfSize = {0, 0};
+        }
+
+        void ShelfNextFitPacker::store(Vec2<PackingSizeType> rectSize) {
+            packing.rects.push_back({{currentShelfSize.x, usedHeight}, rectSize});
+            currentShelfSize.x += rectSize.x;
+            currentShelfSize.y = std::max(currentShelfSize.y, rectSize.y);
+        }
+
+        unsigned int ceilLog2(uint64_t i) {
```

Comment:
Isn't there an even faster implementation using bitshifts or search for the highest-set bit?
The next power of two of `110110` (6 digits) is `1000000` (7 digits)?

Reply:
The standard bitshift version doesn't really work here, because it calculates the next power of two and not the logarithm. There is a method which basically does binary search for the highest set bit (something like [this](https://stackoverflow.com/a/15327567)), and there's also builtins for gcc, clang and msvc, which compile down to a single instruction (count leading zeros). Do you think it is worth to do compiler based intrinsic selection, or is the binary search one fast enough?

Reply:
Binary search method implemented in 701414d